### PR TITLE
[SPARK-40042][PYTHON][DOCS] Make pyspark.sql.streaming.query examples self-contained

### PR DESCRIPTION
**Pull request by dcoliversun**
### What changes were proposed in this pull request?

This PR proposes to improve the examples in `pyspark.sql.streaming.query` by making each example self-contained with a brief explanation and a bit more realistic example.

### Why are the changes needed?

To make the documentation more readable and able to copy and paste directly in PySpark shell.

### Does this PR introduce _any_ user-facing change?

Yes, it changes the documentation.

### How was this patch tested?

Manually ran each doctest.

### CODE REVIEW

File: python/pyspark/sql/streaming/query.py [hunk 1]
```diff
@@ -337,7 +352,6 @@ def _test() -> None:
 
     globs["spark"] = spark
     globs["sqlContext"] = SQLContext.getOrCreate(spark.sparkContext)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:354]: We should probably remove `sqlContext` too (and replace the usages in the doctests to `spark`).

File: python/pyspark/sql/streaming/query.py [hunk 2]
```diff
@@ -165,6 +165,11 @@ def explain(self, extended: bool = False) -> None:
 
         Examples
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:166]: Should also probably add examples to the API that do not have examples, e.g., `stop`.

Comment (HyukjinKwon) [reply python/pyspark/sql/streaming/query.py:166]: or `processAllAvailable`

File: python/pyspark/sql/streaming/query.py [hunk 3]
```diff
@@ -165,6 +165,11 @@ def explain(self, extended: bool = False) -> None:
...
         --------
+        >>> sdf = spark.readStream.format("text").load("python/test_support/sql/streaming")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:168]: I think we can also switch this to `spark.readStream.format("rate").load()`

File: python/pyspark/sql/streaming/query.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/streaming/query.py:354]: Or remove it since SQLContext is deprecated.

**Comment by xinrong-meng** [comment]
Thanks @dcoliversun! We may also want to document parameters, e.g. `id` in `get` misses documents.

Also feel free to open another PR if you prefer :)

File: python/pyspark/sql/streaming/query.py (hunk 1 above)
Comment (dcoliversun) [reply python/pyspark/sql/streaming/query.py:354]: Good catch. Address in next commit :)

**Comment by dcoliversun** [comment]
@xinrong-meng Good idea. I will do this on another PR :)

File: python/pyspark/sql/streaming/query.py (hunk 2 above)
Comment (dcoliversun) [reply python/pyspark/sql/streaming/query.py:295]: Addressed in https://github.com/apache/spark/pull/37482/commits/7d890d68fc82d860d6783976f30673d11d7ef4c9

File: python/pyspark/sql/streaming/query.py (hunk 3 above)
Comment (dcoliversun) [reply python/pyspark/sql/streaming/query.py]: Addressed in https://github.com/apache/spark/commit/7d890d68fc82d860d6783976f30673d11d7ef4c9

**Comment by HyukjinKwon** [comment]
@viirya and @HeartSaVioR FYI

File: python/pyspark/sql/streaming/query.py [hunk 4]
```diff
@@ -316,14 +501,33 @@ def removeListener(self, listener: StreamingQueryListener) -> None:
...
+        ...
+        ...     def onQueryProgress(self, event):
+        ...         pass
+        ...
+        ...     def onQueryTerminated(self, event):
+        ...         pass
+
+        >>> test_listener = TestListener()
+
+        Register streaming query listener
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:518]:
```suggestion
        Register streaming query listener

```

File: python/pyspark/sql/streaming/query.py [hunk 5]
```diff
@@ -316,14 +501,33 @@ def removeListener(self, listener: StreamingQueryListener) -> None:
...
+        >>> spark.streams.addListener(test_listener)
+
+        Deregister streaming query listener
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:521]:
```suggestion
        Deregister streaming query listener

```

File: python/pyspark/sql/streaming/query.py [hunk 6]
```diff
@@ -316,14 +501,33 @@ def removeListener(self, listener: StreamingQueryListener) -> None:
...
+        >>> class TestListener(StreamingQueryListener):
+        ...     def onQueryStarted(self, event):
+        ...         pass
+        ...
+        ...     def onQueryProgress(self, event):
+        ...         pass
+        ...
+        ...     def onQueryTerminated(self, event):
+        ...         pass
+
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:515]:
```suggestion
```

File: python/pyspark/sql/streaming/query.py [hunk 7]
```diff
@@ -300,6 +467,24 @@ def addListener(self, listener: StreamingQueryListener) -> None:
...
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:483]:
```suggestion
```

File: python/pyspark/sql/streaming/query.py [hunk 8]
```diff
@@ -300,6 +467,24 @@ def addListener(self, listener: StreamingQueryListener) -> None:
...
+        >>> test_listener = TestListener()
+
+        Register streaming query listener
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:486]:
```suggestion
        Register streaming query listener

```

File: python/pyspark/sql/streaming/query.py [hunk 9]
```diff
@@ -300,6 +467,24 @@ def addListener(self, listener: StreamingQueryListener) -> None:
...
+        >>> spark.streams.addListener(test_listener)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/query.py:487]: Let's remove the listener back here too to avoid affecting other tests.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/streaming/query.py (hunk 9 above)
Comment (dcoliversun) [reply python/pyspark/sql/streaming/query.py:487]: Thanks, updated :)

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**